ICAFIX/scripts/melodicicasso.m

```diff
@@ -0,0 +1,348 @@
+  vis = 'basic';
+end
+if nargin < 8 || isempty(vis) 
+  vis = 'basic';
+end
+if ~ismember(vis,{'basic','off'})
+  warning('vis must be ''basic'' or ''off'', reverting to ''basic''')
+  vis = 'basic';
+end
+if nargin < 9 || isempty(seedOffset) 
+  seedOffset = 0;
+else
+  seedOffset=str2double(seedOffset);
+end
+if nargin < 10 || isempty(nThreads) 
+  nThreads = min(5,feature('numcores'));
+else
+  nThreads = str2double(nThreads);
+end
+
+%% check IO dependencies
+% use imaging processing toolbox utilities, or if those are not available use FSL utilities 
+function out = out2(fun);[~,out] = fun();end
+function out = out3(fun);[~,~,out] = fun();end
+if isempty(which('niftiread'))
+  if isempty(which('read_avw'))
+    error('neither niftiread nor read_avw on matlab path!')
+  end
+  infoNIFTI = @(fName) struct('ImageSize',out2(@() read_avw(fName))','PixelDimensions',out3(@() read_avw(fName))');
+  readNIFTI = @(fName) read_avw(fName);
+  writeNIFTI = @(img,fName,hdr) save_avw(img,fName,'f',hdr.PixelDimensions);
+else
+  readNIFTI = @(fName) niftiread(fName);
+  infoNIFTI = @(fName) niftiinfo(fName);
+  writeNIFTI = @(img,fName,hdr) niftiwrite(img,fName,hdr,'Compressed',true);
+end
+
+%% parse paths
+warning('off','MATLAB:MKDIR:DirectoryExists')
+% inputs
+vntsFile = sprintf('%s/%s_vnts.nii.gz',ConcatFolder,concatfmrihp);
+brainMaskFile = sprintf('%s/%s_brain_mask.nii.gz',ConcatFolder,concatfmri);
+if ~exist(vntsFile,'file');error('%s doesn''t exist!',vntsFile);end
+if ~exist(brainMaskFile,'file');error('%s doesn''t exist!',brainMaskFile);end
+
+% outputs
+outDir = sprintf('%s/%s.ica/filtered_func_data.ica',ConcatFolder,concatfmrihp);
+if ~mkdir(outDir);error('Unable to make output folder!');end
+warning('on','MATLAB:MKDIR:DirectoryExists')
+
+%% load data
+vnts = double(readNIFTI(vntsFile));% icasso expects double
+hdr = infoNIFTI(vntsFile);
+imSz = size(vnts);
+brainMask = logical(readNIFTI(brainMaskFile));
+[vnts,mtxDim] = maskAndSpatiallyFlatten(vnts,brainMask);
+N = size(vnts,1);
+
+%% run pre-icasso melodics
+init_ica = [];
+nL = numel(nICA);
+for iL = 1:nL
+  nI = nICA(iL);
+  nSteps = zeros(nI,1);
+%   tmpDir = tempname;
+  tmpDir = [outDir '/tmp_melodics'];
+  if exist(tmpDir,'dir');rmdir(tmpDir,'s');end
+  mkdir(tmpDir);
+  [A,W] = deal(cell(1,nI));
+  bootIdx = round(rand(N,nI).*N + 0.5);
+  if iL == 1 && strcmp(bootMode,'randinit')
+    bootFile = sprintf('%s/melodic_boot_vnts',tmpDir);
+    X = single(vnts);
+    X = unmaskAndSpatiallyInflate(X,imSz,brainMask,mtxDim);
+    writeNIFTI(X,bootFile,hdr);
+    fprintf('Running with randinit only (no bootstrapping)\n'); % no bootstrapping
+  else
+    fprintf('Running with bootstrapping\n'); % prepare bootstraps (resampling with replacement)
+  end
+
+  % run parfor over melodics
+  if nThreads > 1
+    if isempty(gcp('nocreate'));parpool(nThreads);end
+    fprintf('Level %i: runing %i melodics in parallel with %i cores:\n',iL,nI,nThreads)
+    parfor iI = 1:nI
```

Comment:
Please restore the edits I made to disable parallel here, so that recompiling doesn't produce a 200MB file.  We don't want to have to remember to do this every time we recompile (and different behavior between interpreted and compiled would be confusing).

Reply:
We need to run melodicicasso in some kind of parallel, as melodic is a single threaded process and melodicicasso would be too slow without parallelization.

Reply:
The compiled version is too big (220MB) when there is any `parfor` or `parpool` stuff in it (even when the entire `parfor` loop body is commented out), and github rejects it from being pushed (though it suggests "lfs" as a way around it, but that size of binary is pretty offensive anyway).  Since the main thing that is in parallel is a `system()` call anyway, can we use bash to launch melodic in parallel, instead of using matlab's parallel to do it?